File: python/tvm/script/builder/tir/var.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""TVM Script TIR Buffer"""
from tvm._ffi import register_object as _register_object
from tvm.ir import Array, PrimExpr, Range
from tvm.runtime import DataType, Object
from tvm.tir import BufferLoad, BufferRegion, IntImm, Var

from . import _ffi_api


@_register_object("script.builder.tir.Buffer")
class Buffer_(Object):
    def __init__(
        self,
        shape,
        dtype="float32",
        name="buffer",
        data=None,
        strides=None,
        elem_offset=None,
        scope="",
        data_alignment=0,
        offset_factor=0,
        buffer_type="",
        axis_separators=None,
    ) -> None:
        self.__init_handle_by_constructor__(
            _ffi_api.Buffer,
            shape,
            dtype,
            name,
            data,
            strides,
            elem_offset,
            scope,
            data_alignment,
            offset_factor,
            buffer_type,
            axis_separators,
        )

    @property
    def data(self) -> Var:
        return self.buffer.data

    @property
    def dtype(self) -> DataType:
        return self.buffer.dtype

    @property
    def shape(self) -> Array:
        return self.buffer.shape

    @property
    def axis_separators(self) -> Array:
        return self.buffer.axis_separators

    @property
    def strides(self) -> Array:
        return self.buffer.strides

    @property
    def elem_offset(self) -> PrimExpr:
        return self.buffer.elem_offset

    @property
    def name(self) -> str:
        return self.buffer.name

    @property
    def data_alignment(self) -> int:
        return self.buffer.data_alignment

    @property
    def offset_factor(self) -> int:
        return self.buffer.offset_factor

    @property
    def buffer_type(self) -> int:
        return self.buffer.buffer_type

    def __getitem__(self, indices):
        if not isinstance(indices, (tuple, list)):
            indices = [indices]
        if any(isinstance(index, slice) for index in indices):
            region = []
            for index in indices:
                if isinstance(index, slice):
                    region.append(Range(index.start, index.stop))
                else:
                    region.append(Range.from_min_extent(index, 1))
            return BufferRegion(self.buffer, region)
        else:
            return BufferLoad(self.buffer, indices)


class BufferProxy:
    def __call__(
        self,
        shape,
        dtype="float32",
        name="buffer",
        data=None,
        strides=None,
        elem_offset=None,
        scope="",
        data_alignment=0,
        offset_factor=0,
        buffer_type="",
        axis_separators=None,
    ) -> Buffer_:
        return Buffer_(
            shape,
            dtype,
            name,
            data,
            strides,
            elem_offset,
            scope,
            data_alignment,
            offset_factor,
            buffer_type,
            axis_separators,
        )

    def __getitem__(self, keys) -> Buffer_:
        return self(*keys)  # pylint: disable=no-member # type: ignore


Buffer = BufferProxy()

File: src/script/builder/tir/var.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "./var.h"

namespace tvm {
namespace script {
namespace builder {
namespace tir {

Buffer::Buffer(Array<PrimExpr> shape, DataType dtype, String buffer_name,
               Optional<tvm::tir::Var> data, Optional<Array<PrimExpr>> strides,
               Optional<PrimExpr> elem_offset, String storage_scope, int align, int offset_factor,
               String buffer_type_str, Optional<Array<IntImm>> axis_separators) {
  ObjectPtr<BufferNode> n = make_object<BufferNode>();
  tvm::tir::Var buffer_data;
  if (!data.defined()) {
    DataType storage_dtype = dtype;
    if (storage_dtype == DataType::Bool()) {
      storage_dtype = DataType::Int(8);
    }
    buffer_data = tvm::tir::Var(buffer_name, PointerType(PrimType(storage_dtype), storage_scope));
  } else {
    buffer_data = data.value();
  }
  tvm::tir::BufferType buffer_type =
      (buffer_type_str == "auto_broadcast") ? tvm::tir::kAutoBroadcast : tvm::tir::kDefault;
  n->buffer = tvm::tir::Buffer(buffer_data, dtype, shape, strides.value_or(Array<PrimExpr>()),
                               elem_offset.value_or(PrimExpr()), buffer_name, align, offset_factor,
                               buffer_type, axis_separators.value_or(Array<IntImm>()));
  data_ = n;
}

tvm::tir::BufferLoad BufferNode::BufferLoad(Array<PrimExpr> indices) {
  return tvm::tir::BufferLoad(buffer, indices);
}

tvm::tir::BufferStore BufferNode::BufferStore(PrimExpr value, Array<PrimExpr> indices) {
  return tvm::tir::BufferStore(buffer, value, indices);
}

tvm::tir::BufferRegion BufferNode::BufferRegion(Array<Range> region) {
  return tvm::tir::BufferRegion(buffer, region);
}

tvm::tir::Prefetch BufferNode::Prefetch(Array<Range> bounds) {
  return tvm::tir::Prefetch(buffer, bounds);
}

TVM_STATIC_IR_FUNCTOR(Namer, vtable)
    .set_dispatch<tvm::tir::BufferNode>([](const ObjectRef& node, String name) -> void {
      tvm::tir::BufferNode* buffer =
          const_cast<tvm::tir::BufferNode*>(node.as<tvm::tir::BufferNode>());
      buffer->name = name;
      Namer::Name(buffer->data, name + "_data");
      int n = buffer->strides.size();
      for (int i = 0; i < n; ++i) {
        PrimExpr e = buffer->strides[i];
        if (const tvm::tir::VarNode* v = e.as<tvm::tir::VarNode>()) {
          Namer::Name(GetRef<tvm::tir::Var>(v), name + "_s" + std::to_string(i));
        }
      }
    });

TVM_STATIC_IR_FUNCTOR(Namer, vtable)
    .set_dispatch<BufferNode>([](const ObjectRef& node, String name) -> void {
      Namer::Name(node.as<BufferNode>()->buffer, name);
    });

TVM_STATIC_IR_FUNCTOR(Namer, vtable)
    .set_dispatch<tvm::tir::SizeVarNode>([](const ObjectRef& node, String name) -> void {
      using namespace tvm::tir;
      SizeVarNode* var = const_cast<SizeVarNode*>(node.as<SizeVarNode>());
      var->name_hint = name;
    });

TVM_STATIC_IR_FUNCTOR(Namer, vtable)
    .set_dispatch<tvm::tir::VarNode>([](const ObjectRef& node, String name) -> void {
      using namespace tvm::tir;
      VarNode* var = const_cast<VarNode*>(node.as<VarNode>());
      var->name_hint = name;
    });

TVM_STATIC_IR_FUNCTOR(Namer, vtable)
    .set_dispatch<tvm::tir::IterVarNode>([](const ObjectRef& node, String name) -> void {
      using namespace tvm::tir;
      IterVarNode* var = const_cast<IterVarNode*>(node.as<IterVarNode>());
      Namer::Name(var->var, name);
    });

TVM_REGISTER_NODE_TYPE(BufferNode);
TVM_REGISTER_GLOBAL("script.builder.tir.Buffer")
    .set_body_typed([](Array<PrimExpr> shape, DataType dtype = DataType::Float(32),
                       String buffer_name = "buffer", Optional<tvm::tir::Var> data = NullOpt,
                       Optional<Array<PrimExpr>> strides = NullOpt,
                       Optional<PrimExpr> elem_offset = NullOpt, String storage_scope = "",
                       int align = 0, int offset_factor = 0, String buffer_type_str = "",
                       Optional<Array<IntImm>> axis_separators = NullOpt) {
      return Buffer(shape, dtype, buffer_name, data, strides, elem_offset, storage_scope, align,
                    offset_factor, buffer_type_str, axis_separators);
    });

TVM_REGISTER_GLOBAL("script.builder.tir.BufferBufferLoad")
    .set_body_method<Buffer>(&BufferNode::BufferLoad);
TVM_REGISTER_GLOBAL("script.builder.tir.BufferBufferRegion")
    .set_body_method<Buffer>(&BufferNode::BufferRegion);

}  // namespace tir
}  // namespace builder
}  // namespace script
}  // namespace tvm
